### PR TITLE
Enhancement: Add parsers and serializers for JSON, YAML and XML

### PR DESCRIPTION
**Pull request by zykure**
This PR adds parsing and serialization features for structured data in JSON, YAML and XML formats.

For each format, a new function `to_*_string()` and `from_*_string()` is exposed in the gul17 library via the `data_processors.h` header. These functions convert between the string representation of the data and the new `DataTree` structure, which contains the corresponding tree nodes.

The `DataTree` struct is essentially a wrapper around an `std::variant` value that can hold null, boolean, integer, float and string values as well as other `DataTree` objects in the form of a map or vector. This allows to represent each data item by a node in the tree, and includes sub-trees and node arrays.

The XML parser/serializer also handles attributes and multiple elements of the same name. Attributes are added as tree nodes with a name like `"@attr"` under the current XML element, and multiple elements that have the same XML tag on the same tree level are added as a node array. This is not a full-scale DOM/SAX-style parser of course, so there are some limitations.

Unit tests for the 3 formats are provided as well.

I think this enhancement would be useful to other projects such as the DOOCS serverlib/clientlib where XML is already being used; and JSON/YAML parsing would add a new and convenient way of transferring structured data via the DOOCS protocol or to read and write configuration files, etc.

### CODE REVIEW

File: include/gul17/DataTree.h [hunk 1]
```diff
@@ -0,0 +1,656 @@
...
+        Array,             // array
+        Object             // object
+    >;
+
+    // Constructors
+
+    /**
+     * Create an empty DataTree object (default to an empty object).
+     */
+    DataTree() : value_(Object()) {}  // default to an empty object
```
[the remark this answers is not in the document]
Comment (zykure) [reply include/gul17/DataTree.h:90]: No, I believe this is correct. Accessing a non-existent key should create a new empty object, so that new tree nodes can be created easily:

```cpp
DataTree tree;
tree["key1"]["key2"]["key3"] = "value";
```

**Comment by Finii** [comment]
This AI cruft does not make the PR human readable. Digging through at this moment already 68 comments is nothing I like and have time for. Maybe you should do the AI processing in a private chat or what and keep this for humans.

Anyhow, I did not look at the code at all, but
* I believe we should use one of the several available and thoroughly tested libraries for parsing and not invent parsers on our own.
* And with use I mean in the other libraries / servers; not pulling these libs in into GUL
* We use Xerces for example in a lot libraries and servers, why switch away from a well maintained library?

> and JSON/YAML parsing would add a new and convenient way of transferring structured data via the DOOCS protocol

Iirc @soerengrunewald did a lot research on the serialization and it is always best to use one of the existing libraries and not invent the wheel again.
Yaml for example has some severe quoting challenges, and I'm not sure we are able to catch them all - and also see no need to add a maintenance burden on us instead of relying on widespread used libraries.

_Edit: Add last paragraph_

**Comment by Finii** [comment]
My little Copilot chat...

> Is it better to use the xerces library to manipulate XML data in a c++ program, or shall we write our own parser? Please give extensive reasoning for you answer.

Using an Existing Library vs. Rolling Your Own Parser

When it comes to parsing and manipulating XML data in C++, there are two primary options: using an existing library like Xerces or writing your own parser from scratch. Here's a detailed analysis of the pros and cons of each approach:
Using the Xerces Library

Pros:

* Maturity and Stability: Xerces is a well-established, widely-used, and mature library with a large user base. This translates to a more stable and reliable parsing experience.
* Standards Compliance: Xerces is designed to be compliant with various XML standards, including XML 1.0, XML 1.1, and XML Schema.
* Feature-Rich: Xerces provides a comprehensive set of features, including support for DOM, SAX, and StAX parsing, as well as XML validation, XPath, and XSLT.
* Community Support: As a popular library, Xerces has an active community, which means there are many resources available for troubleshooting and learning.
* Less Code to Maintain: By using an existing library, you can focus on your application's logic without worrying about the intricacies of XML parsing.

Cons:

* Overhead: Using a library like Xerces can introduce additional overhead, including increased binary size and potential performance impacts.
* Learning Curve: While Xerces is well-documented, it still requires time and effort to learn how to use it effectively.

Writing Your Own Parser

Pros:

* Customizability: By writing your own parser, you can tailor it to your specific needs and optimize it for your use case.
* Lightweight: A custom parser can be designed to be lightweight, reducing binary size and overhead.
* Fine-Grained Control: With a custom parser, you have complete control over the parsing process, allowing for precise error handling and optimization.

Cons:

* Complexity: Writing a robust and standards-compliant XML parser is a complex task, requiring significant expertise and time.
* Bugs and Edge Cases: A custom parser is more likely to contain bugs and may not handle edge cases as well as a mature library like Xerces.
* Maintenance Burden: Your team will need to invest time and resources in maintaining and updating the parser, which can be a significant burden.

Conclusion

Unless you have a specific requirement that cannot be met by an existing library like Xerces (e.g., a highly customized or embedded system), it is generally recommended to use an established library like Xerces for XML parsing and manipulation in C++. The benefits of using a mature library, including stability, standards compliance, and community support, outweigh the potential drawbacks.

Writing your own parser should only be considered if:

* You have a very specific, lightweight, or embedded system requirement.
* You have extensive experience with XML parsing and are confident in your ability to create a robust and standards-compliant parser.

In most cases, using Xerces or another established library will save you time, effort, and potential headaches in the long run.

**Review by alt-graph — COMMENTED** [review]
This looks useful to me, so it might tick the "Only code that is potentially useful for a wide range of C++ projects" box. What worries me a little is the lack of usage experience.

@zykure: Have you already used this or similar code either in multiple projects or intensively in a single project? If not, it might be better to put it into a separate library and gather experience there. Once it gets into GUL we'll be **very** unwilling to change the API again.

Do we have other project where we would want to use this? The clientlib already has `Structure` which is similar but different.

Other observations from a casual glance at the code:
- "Data processors" might be a misnomer. Maybe "parser" is closer to what the code is actually doing?
- The free functions `from_json_string()` etc. do not have enough context for their names. It might be better to embed them into the `DataTree` class as static functions, e.g. `auto tree = DataTree::from_json_string()`.

**Comment by Finii** [comment]
I'm not sure how other people do it, but in our servers that use XML they _**of course**_ validate the input XML via XSD.
That should of course be part of "our parser" if we have one, as good practice would be to validate the data before using it for better error handling and reporting.
Is that also included here? Without validation we would encourage people to use XML un-validated. The XSD data that belongs to a XML input also help documenting what we really expect and allow, giving a (automatically validated) protocol description.

**Comment by Finii** [comment]
@alt-graph wrote:

> it might be better to put it into a separate library and gather experience there

I see no reason why it should/could not stay in a separate library. :thinking:
Do not really see this fitting into GUL. 

> tick the "Only code that is potentially useful for a wide range of C++ projects" box

Well, but otoh we need to keep GUL "small enough" and not become the bucket for everything.
We've had other functionality not added to GUL, like Ryu, like `fmt`, like gsl, ...
I vote for a separate library (or using external libraries anyhow).

**Comment by soerengrunewald** [comment]
I think @Finii is on to something. I would rather see GSL features in GUL over some parsers.

Also there are battle tested parsers out there (which fixed all the edge cases), may of them header only and available via wrapdb, e.g:
* https://nlohmann.github.io/json/
* https://marzer.github.io/tomlplusplus/
* https://github.com/jbeder/yaml-cpp

That been said, I also see the benefit of having a unified interfaces for any type of parser. Non the less I don't see it gul.

On the code side there is also some things which I would like to see changed.
- Generally the API seems inconsistend.
- `DataTree::at` `DataTree::operator[]` key access should be `string_views`
- `DataTree::as` `DataTree::is` seem to differ in behavior, might be wanted, but seems strange.
- `const_cast` not a fan.
- General style issues (e.g large if-else blocks and loops, deep nesting, hard coded strings/numerical values)

**Comment by alt-graph** [comment]
Thanks a lot for the PR, @zykure. I think this would make a very useful stand-alone library. Overall, however, a majority seems to feel that this addition is too "high-level" to be a good fit for GUL17. I'll close the PR.
